pkg/utils/logger/zaplogger.go

```diff
@@ -0,0 +1,173 @@
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package logger
+
+import (
+	"os"
+	"strings"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	lumberjack "gopkg.in/natefinch/lumberjack.v2"
+)
+
+type structuredLogger struct {
+	zapLogger *zap.SugaredLogger
+}
+
+// getZapLevel converts log level string to zapcore.Level.
+func getZapLevel(logLevel string) zapcore.Level {
+	switch lvl := strings.ToLower(os.Getenv(envLogLevel)); lvl {
+	case "debug":
+		return zapcore.DebugLevel
+	case "info":
+		return zapcore.InfoLevel
+	case "warn":
+		return zapcore.WarnLevel
+	case "error":
+		return zapcore.ErrorLevel
+	case "fatal":
+		return zapcore.FatalLevel
+	default:
+		return zapcore.DebugLevel
+	}
+}
+
+func (l *structuredLogger) Debugf(format string, args ...interface{}) {
+	l.zapLogger.Debugf(format, args...)
+}
+
+func (l *structuredLogger) Debug(format string) {
+	l.zapLogger.Desugar().Debug(format)
+}
+
+func (l *structuredLogger) Infof(format string, args ...interface{}) {
+	l.zapLogger.Infof(format, args...)
+}
+
+func (l *structuredLogger) Info(format string) {
+	l.zapLogger.Desugar().Info(format)
+}
+
+func (l *structuredLogger) Warnf(format string, args ...interface{}) {
+	l.zapLogger.Warnf(format, args...)
+}
+
+func (l *structuredLogger) Warn(format string) {
+	l.zapLogger.Desugar().Warn(format)
+}
+
+func (l *structuredLogger) Error(format string) {
+	l.zapLogger.Desugar().Error(format)
+}
+
+func (l *structuredLogger) Errorf(format string, args ...interface{}) {
+	l.zapLogger.Errorf(format, args...)
+}
+
+func (l *structuredLogger) Fatalf(format string, args ...interface{}) {
+	l.zapLogger.Fatalf(format, args...)
+}
+
+func (l *structuredLogger) Panicf(format string, args ...interface{}) {
+	l.zapLogger.Fatalf(format, args...)
+}
+
+func (l *structuredLogger) WithFields(fields Fields) Logger {
+	var f = make([]interface{}, 0)
+	for k, v := range fields {
+		f = append(f, k)
+		f = append(f, v)
+	}
+	newLogger := l.zapLogger.With(f...)
+	return &structuredLogger{newLogger}
+}
+
+func getEncoder() zapcore.Encoder {
+	encoderConfig := zap.NewProductionEncoderConfig()
+	encoderConfig.EncodeTime = zapcore.ISO8601TimeEncoder
+	return zapcore.NewJSONEncoder(encoderConfig)
+}
+
+// GetLogFileLocation returns the log file path
+func getLogFileLocation(srvName string) string {
+	switch srvName {
+	case "L-IPamD":
+		logFilePath := os.Getenv(envIPamDLogFilePath)
+		if logFilePath == "" {
+			logFilePath = defaultIPamDLogFilePath
+		}
+		return logFilePath
+	case "Plugin":
+		logFilePath := os.Getenv(envPluginFilePath)
+		if logFilePath == "" {
+			logFilePath = defaultPluginLogFilePath
+		}
+		return logFilePath
+	default:
+		return "stdout"
+	}
+}
+
+func newZapLogger(srvName string) Logger {
```

Comment:
Does `srvName` mean "server name"? "service name"? Looking at the code, it seems to mean "component", would that be a better name?

Reply:
`srvName` means server name. The variable is used to differentiate between the log file paths for different binaries which we have. For example, if the server is `L-IPamD`, it written to `/host/var/log/aws-routed-eni/ipamd.log` and if the the server is `Plugin`, the logs are written to `/var/log/aws-routed-eni/plugin.log`.

Reply:
I got how it's used, I was just confused that the plugin was called a server. It's a binary that gets executed 😄